csrc/runtime/fusion_kernel_runtime.cpp

```diff
@@ -330,6 +330,45 @@ KernelArgumentHolder FusionKernelRuntime::runWithInputs(
   return fusion_outputs;
 }
 
+std::vector<KernelArgumentHolder> FusionKernelRuntime::prepareInputs(
+    const KernelArgumentHolder& args) const {
+  std::vector<KernelArgumentHolder> all_runtime_inputs;
+  const int64_t num_groups = numGroups();
+  all_runtime_inputs.reserve(num_groups);
+
+  ArgumentManager args_manager(
+      args, runtime_workspace_, segmented_fusion_->inputs());
+
+  // group should share cache id.
+  const auto group_cache_id = args.getCacheId();
+
+  for (int64_t run_order_id = 0; run_order_id < num_groups; ++run_order_id) {
+    auto group_to_run = runtime_workspace_.group_run_order.at(run_order_id);
+    // TODO: index mode should be updated per segmented kernel
+    // Prepare input vector
+    all_runtime_inputs.push_back(
+        args_manager.translateValsToArgs(group_to_run->inputs()));
+    auto& group_runtime_inputs = all_runtime_inputs.back();
+
+    group_runtime_inputs.setDeviceIndex(args.getDeviceIndex());
+    if (group_cache_id.has_value()) {
+      group_runtime_inputs.setCacheId(group_cache_id.value());
+    }
+
+    // A segment fusion is created again in compileKernel. Can be done
+    // just once?
```

Comment:
(Can be done as a follow-up)

I don't think we even need to create a segment fusion here. https://github.com/NVIDIA/Fuser/blob/87365adb01b25df8160130a19ca704e8f38c20a3/csrc/runtime/allocations.cpp#L50-L54 can be computed based on the complete fusion just fine.

Reply:
I think you're right. I'll leave it as is since it doesn't seem like a small change.